File: 624/Solution.java
class Solution {
    public int maxDistance(List<List<Integer>> arrays) {
        int min = arrays.get(0).getFirst();
        int max = arrays.get(0).getLast();
        int res = 0;
        for(int i = 1; i < arrays.size(); i++){
            int curMin = arrays.get(i).getFirst();
            int curMax = arrays.get(i).getLast();
            res = Math.max(Math.max(curMax - min, max - curMin), res);
            min = Math.min(min, curMin);
            max = Math.max(max, curMax);
        }
        return res;
    }
}
// time O(n)
// space O(1)

File: 624/Solution.go
func maxDistance(a [][]int) int {
    min, max := a[0][0], a[0][len(a[0]) - 1]
    res := 0
    for i  := 1; i < len(a); i++{
        curMin, curMax := a[i][0], a[i][len(a[i]) - 1]
        res = maxInt(res, maxInt(curMax - min, max - curMin))
        if min > curMin { min = curMin }
        if max < curMax { max = curMax }
    }
    return res
}
func maxInt(x, y int) int {
    if x > y {
        return x
    }
    return y
}
